module/zfs/dmu_objset.c

```diff
@@ -1639,20 +1640,67 @@ sync_dnodes_task(void *arg)
 
 	multilist_sublist_unlock(ms);
 
+	zio_nowait(sda->sda_cio);
+
 	kmem_free(sda, sizeof (*sda));
 }
 
+typedef struct sync_objset_arg {
+	objset_t *soa_os;
+	dmu_tx_t *soa_tx;
+	zio_t *soa_zio;
+	taskq_ent_t soa_tq_ent;
+} sync_objset_arg_t;
+
+static void
+sync_dnodes_finish_task(void *arg)
+{
+	sync_objset_arg_t *soa = arg;
+	objset_t *os = soa->soa_os;
+	dmu_tx_t *tx = soa->soa_tx;
+
+	/* Free intent log blocks up to this tx. (May block.) */
+	zil_sync(os->os_zil, tx);
+	os->os_phys->os_zil_header = os->os_zil_header;
+
+	zio_nowait(soa->soa_zio);
+	kmem_free(soa, sizeof (*soa));
+}
 
-/* called from dsl */
 void
-dmu_objset_sync(objset_t *os, zio_t *pio, dmu_tx_t *tx)
+dmu_objset_sync_sublists_done(zio_t *zio)
+{
+	sync_objset_arg_t *soa = zio->io_private;
+	objset_t *os = soa->soa_os;
+	dmu_tx_t *tx = soa->soa_tx;
+	int txgoff = tx->tx_txg & TXG_MASK;
+	dbuf_dirty_record_t *dr;
+
+	list_t *list = &DMU_META_DNODE(os)->dn_dirty_records[txgoff];
+	while ((dr = list_remove_head(list)) != NULL) {
```

Comment:
Can this list be manipulated by multiple threads now?

Reply:
The dmu_objset_sync_sublists_done() is called when the sio has reached the _done stage, so each of the sio's child zios should also have already reached the _done stage.  There should be no other threads operating on that os at the point where  dmu_objset_sync_sublists_done() is invoked.

(This is comparable to what the previous code had done, after invoking the taskq_wait(). Basically, it still happens once per os.)

Reply:
Let's and an ASSERT for this to be sure.

Reply:
code updated and ASSERT included.